File: cpp/leetcode/128.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> map(nums.begin(), nums.end());
        int res = 0;
        for(auto num : nums){
            if(map.find(num) == map.end()) continue;
            map.erase(num);
            int l = num - 1, r = num + 1;
            while(map.find(l) != map.end()) map.erase(l--);
            while(map.find(r) != map.end()) map.erase(r++);
            res = max(res, r - l - 1);
        }
        return res;
    }
};


class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_map<int, int> map;
        int res = 0;
        for(int i = 0; i < nums.size(); i++){
            if(map.find(nums[i]) == map.end()){
                int l = map.find(nums[i] - 1) == map.end() ? 0 : map[nums[i] - 1];
                int r = map.find(nums[i] + 1) == map.end() ? 0 : map[nums[i] + 1];
                map[nums[i]] = l + r + 1;
                map[nums[i] - l] = l + r + 1;
                map[nums[i] + r] = l + r + 1;
                res = max(res, l +  r + 1);
            }
        }
        return res;
    }
};

File: python/leetcode/128.py
class Solution:
    def longestConsecutive(self, nums):
        nums = set(nums)
        ans = 0
        for num in nums:
            if num - 1 not in nums:
                m = num + 1
                while m in nums:
                    m += 1
                ans = max(ans, m - num)
        return ans



class Solution:
    def longestConsecutive(self, nums):
        res = 1
        ans = {}
        for n in nums:
            if n not in ans:
                left = ans.get(n-1, 0)
                right = ans.get(n+1, 0)
                total = left + right + 1
                ans[n] = total
                res = max(res, total)
                ans[n - left] = total
                ans[n + right] = total
        return res

